File: frontend/src/lib/permissions.ts
/**
 * The one place onboarding and the setup check both go to ask about a
 * permission, and the one place that opens a macOS privacy pane for either of
 * them.
 *
 * Before this module existed, the two onboarding call sites and the setup
 * check each built their own `invoke('open_system_settings', ...)` call, and
 * one of them left the argument off entirely (FR-04): Tauri's argument
 * deserialisation failed, the promise rejected, and the user landed on a raw
 * `window.alert()` instead of System Settings. Routing every caller through
 * `openPermissionSettings` means there is exactly one call to get right.
 *
 * The permission checks here are evidence-based on the Rust side —
 * `trigger_microphone_permission` and `trigger_system_audio_permission_command`
 * open the device and wait to actually observe an audio callback before they
 * report anything back (see FR-02/FR-03 in
 * docs/audits/2026-09-21-first-run-flow.md). This module does not add any
 * verification of its own; it only relays whatever `PermissionReport` the
 * backend resolved with, or turns a failed `invoke` into an `undetermined`
 * result so a broken Tauri call reads as "we don't know" rather than as a
 * fabricated "denied".
 */

import { invoke } from '@tauri-apps/api/core';
import type { PermissionReport, PermissionStatus, PermissionVerdict } from '@/types/onboarding';

/// The pane names open_system_settings understands, in one place.
export const PERMISSION_PANES = {
  microphone: 'Privacy_Microphone',
  systemAudio: 'Privacy_ScreenCapture',
} as const;

export type PermissionPane = (typeof PERMISSION_PANES)[keyof typeof PERMISSION_PANES];

/**
 * Mirrors `micCheck.ts`'s own `isMacOS()`: a dynamic import of
 * `@tauri-apps/plugin-os` so a non-Tauri context (tests, a stray web render)
 * does not crash, falling back to a userAgent sniff when the plugin itself is
 * unavailable. `open_system_settings` exists only on macOS, so every caller
 * that might invoke it needs this same check.
 */
export async function isMacOS(): Promise<boolean> {
  try {
    const { platform } = await import('@tauri-apps/plugin-os');
    return platform() === 'macos';
  } catch {
    return typeof navigator !== 'undefined' && navigator.userAgent.includes('Mac');
  }
}

/**
 * Opens a macOS privacy pane. Resolves false when it could not be opened —
 * off macOS, where there is nothing to open, or when the invoke itself
 * failed — so the caller can fall back to written instructions instead of
 * throwing in front of the user.
 *
 * The `{ preferencePane: pane }` argument object is the whole point of this
 * function: `open_system_settings` requires it, and a caller that builds the
 * invoke by hand is exactly how FR-04 happened the first time.
 */
export async function openPermissionSettings(pane: PermissionPane): Promise<boolean> {
  if (!(await isMacOS())) return false;

  try {
    await invoke('open_system_settings', { preferencePane: pane });
    return true;
  } catch (error) {
    console.error('Failed to open system settings:', error);
    return false;
  }
}

/**
 * Turns a failed invoke into the report shape the caller still needs. A
 * command that never reached the backend has told us nothing about the
 * permission itself, so the verdict is `undetermined`, not `denied` — the
 * same reasoning `startSetupCheck` uses for `command_failed`.
 */
function commandFailedReport(error: unknown): PermissionReport {
  console.error('Permission check could not be started:', error);
  const detail = error instanceof Error ? error.message : String(error);
  return { verdict: 'undetermined', detail };
}

/**
 * Asks the backend to verify the microphone by observing audio, and relays
 * whatever `PermissionReport` it resolves with. This function does not
 * itself verify anything — the evidence, such as it is, comes from
 * `trigger_microphone_permission` waiting on an audio callback.
 */
export async function checkMicrophonePermission(): Promise<PermissionReport> {
  try {
    return await invoke<PermissionReport>('trigger_microphone_permission');
  } catch (error) {
    return commandFailedReport(error);
  }
}

/**
 * Asks the backend to verify system audio the same way. `undetermined` here
 * is also exactly what a Mac with nothing playing produces — silence is not
 * evidence of a denial, so the backend does not report one, and neither does
 * this function.
 */
export async function checkSystemAudioPermission(): Promise<PermissionReport> {
  try {
    return await invoke<PermissionReport>('trigger_system_audio_permission_command');
  } catch (error) {
    return commandFailedReport(error);
  }
}

/**
 * The UI row state a wire verdict maps to. Pulled out to its own function —
 * rather than inlined at each of the row's call sites — so the mapping from
 * `PermissionVerdict` to `PermissionStatus` exists exactly once and cannot
 * drift between the microphone row and the system-audio row.
 */
export function verdictToStatus(verdict: PermissionVerdict): PermissionStatus {
  switch (verdict) {
    case 'authorized':
      return 'authorized';
    case 'denied':
      return 'denied';
    case 'undetermined':
      return 'undetermined';
  }
}

File: frontend/src-tauri/src/audio/permissions.rs
// macOS audio permissions handling
//
// What is left in this file opens a settings page and logs where the Audio
// Capture row lives. The one question it used to answer for itself — whether
// system audio is actually reaching the app — now belongs to
// `permission_check.rs`, which answers it from samples that arrived rather than
// from a tap that was constructed. Building a tap was never evidence: macOS
// hands a denied app a tap that opens cleanly and then yields silence, so the
// old check here passed for a user who would get nothing.

use anyhow::Result;
// Only the macOS paths log: off macOS these functions have nothing to say.
#[cfg(target_os = "macos")]
use log::{error, info};

use crate::audio::permission_check::{verify_system_audio, PermissionReport};

#[cfg(target_os = "macos")]
use std::process::Command;

/// Log where the Audio Capture permission (macOS 14.4+) lives. This reports
/// nothing about whether it was granted.
///
/// No reading is taken: it returns `true` on a Mac that has granted Audio
/// Capture and on one that has refused it. macOS exposes no way to query that
/// grant ahead of use — the prompt is raised by Core Audio itself, when
/// `AudioHardwareCreateProcessTap` builds the first tap — so the only honest
/// source of an answer is audio arriving, which
/// `permission_check::verify_system_audio` waits for. Callers that need to know
/// must ask that instead of reading anything into this `true`.
#[cfg(target_os = "macos")]
pub fn check_screen_recording_permission() -> bool {
    info!("ℹ️  Core Audio tap requires Audio Capture permission (macOS 14.4+)");
    info!("📍 The permission dialog appears when the first Core Audio tap is created");
    info!("   To review it: System Settings → Privacy & Security → Audio Capture");

    // Unconditional, for the reason given above: nothing here has been checked.
    true
}

/// The same non-answer off macOS, where no such permission exists.
#[cfg(not(target_os = "macos"))]
pub fn check_screen_recording_permission() -> bool {
    true // No Audio Capture permission on this platform, so nothing to report
}

/// Open Privacy & Security so the user can find the Audio Capture row.
///
/// Opening the page is the whole of it. macOS offers no URL that deep-links to
/// the Audio Capture pane and no API that asks for the grant directly, so the
/// user has to find the row themselves, and whether they then grant it is not
/// visible from here.
#[cfg(target_os = "macos")]
pub fn request_screen_recording_permission() -> Result<()> {
    info!("🔐 Opening System Settings for Audio Capture permission...");

    // Open System Settings to Privacy & Security page
    // Note: There's no direct URL for Audio Capture, so we open the main Privacy page
    let result = Command::new("open")
        .arg("x-apple.systempreferences:com.apple.preference.security")
        .spawn();

    match result {
        Ok(_) => {
            // `open` was launched. Whether System Settings came to the front,
            // and what the user does once it is there, is not observed here.
            info!(
                "✅ Asked macOS to open System Settings - navigate to Privacy & Security → \
                 Audio Capture"
            );
            info!("👉 Enable Audio Capture there, then run the system audio check again");
            Ok(())
        }
        Err(e) => {
            error!("❌ Failed to open System Settings: {}", e);
            Err(anyhow::anyhow!("Failed to open System Settings: {}", e))
        }
    }
}

#[cfg(not(target_os = "macos"))]
pub fn request_screen_recording_permission() -> Result<()> {
    Ok(()) // Not required on other platforms
}

/// Tauri command to log where the Screen Recording (Audio Capture) setting is.
///
/// Registered in lib.rs, which is the only reason it still takes this shape;
/// no caller in this repository invokes it today. Its `true` means "logged",
/// not "granted" — see [`check_screen_recording_permission`].
#[tauri::command]
pub async fn check_screen_recording_permission_command() -> bool {
    check_screen_recording_permission()
}

/// Tauri command to open the settings page for the Screen Recording (Audio
/// Capture) permission. Returning `Ok` means the page was asked to open,
/// nothing about the grant.
#[tauri::command]
pub async fn request_screen_recording_permission_command() -> Result<(), String> {
    request_screen_recording_permission().map_err(|e| e.to_string())
}

/// Tauri command behind onboarding's system audio row.
///
/// It decides nothing itself, deliberately: onboarding and the start of a
/// recording both ask `verify_system_audio`, so the two surfaces cannot reach
/// different conclusions about the same tap — which is exactly what they used
/// to do, onboarding passing a tap that recording would find silent. Creating
/// that tap is still what raises the macOS prompt, so the prompt and the
/// verification remain one act; it is the verdict that now comes from samples.
///
/// The `Result` exists for the frontend's invoke signature. A check that could
/// not be made comes back as an `Undetermined` verdict rather than an error, so
/// nothing here returns `Err`.
#[tauri::command]
pub async fn trigger_system_audio_permission_command() -> Result<PermissionReport, String> {
    Ok(verify_system_audio().await)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_screen_recording_check_answers_the_same_on_every_machine() {
        // It returns true whether or not Audio Capture was granted, which is
        // why no caller may read it as evidence of a grant. Pinned here so a
        // change that gives it a real reading has to come through this test and
        // say what it now checks.
        assert!(check_screen_recording_permission());
    }

    /// Off macOS the app captures no system audio at all, so onboarding must
    /// not be handed a grant for it. The macOS path is left to a machine with a
    /// tap: CI runs `cargo test -p meetily` on headless Ubuntu and nothing here
    /// may open a device.
    #[cfg(not(target_os = "macos"))]
    #[tokio::test]
    async fn the_onboarding_command_claims_no_grant_off_macos() {
        use crate::audio::permission_check::PermissionVerdict;

        let report = trigger_system_audio_permission_command()
            .await
            .expect("the system audio check reports a verdict rather than an error");

        assert_eq!(report.verdict, PermissionVerdict::Undetermined);
        assert!(!report.is_authorized());
        assert!(!report.is_denied());
    }
}
